### DISCUSSION TITLE
dynamic branch across fixed number of branches

### DISCUSSION THREAD

**Discussion by gofford** [ropensci/targets#577] [Help]
For dynamic branching, is it possible to split the pattern vector into equal sized chunks for a fixed number of branches? E.g., if I currently `map` over a length 10 pattern, I will have 10 branches. Each of these branches will incur its own IO overheads, and this is not necessarily optimal for jobs with large network usage, etc.. Another option would be to split that length 10 pattern over 2 branches, such that each branch runs 5 tasks. 

Is this possible? 

This is related to `tar_rep` in `tarchetypes`, but `tar_rep` is geared towards repeated runs of a single function to, e.g., run a bootstrap analysis or sampling of a distribution. What I am looking at is having a fixed-length input, shared across a fixed number of branches. 

edit:

In a way, I guess this would be a list-equivalent to `tar_group_*()` functions in `tarchetyes`. So instead of using `tar_group_size()` to group a `data.frame` into batches of a fixed size, there is a way to `slice()` a list into either (a) a number of sub-lists of a fixed size, or (b) a fixed number of sub-lists of variable size.

**Comment by wlandau** [comment]
What you are describing is batching, and I do recommend it for the reasons you identify: https://books.ropensci.org/targets/dynamic.html#batching. The existing utilities are not totally universal/generalized because extending them to completion would get messy, but if you return the right data frame in a `tar_group_size()` or `tar_group_count()` target, you can get the same effect. Instead of a list, you would have a data frame with a list column.

> "Everything is a tibble."
>     - Kirill Müller

**Reply by gofford** [reply]
That does make sense, and I had explored that as a workaround but it seemed clunky. It does work, though, so that's a plus!

Current method is to move the indexed list to a `tibble` called `index_df`, group that with `tar_group_size()`, and then `purrr::map` over the `index_df$index` elements within branches invoked by a `pattern = map(...)`. 

With the `tar_group_*` functions, is the onus on the user to them combine the various branches after they have been processed with `tar_combine()`, or is that handled as part of the `tar_group_*` operation?

**Reply by wlandau** [reply]
> and then purrr::map over the index_df$index elements within branches invoked by a pattern = map(...).

That's the tricky part on my end. It is easy to define batches at the beginning, but it is much harder for downstream dynamic branching targets to know what to do with the branches in the general case. I have found it helpful to handle this on a case by case basis, e.g. `stantargets::tar_stan_mcmc_rep_summary()`.

> With the tar_group_* functions, is the onus on the user to them combine the various branches after they have been processed with tar_combine(), or is that handled as part of the tar_group_* operation?

As with anything implemented exclusively with dynamic branching and no static branching, the `tarchetypes::tar_group_*()` functions perform aggregation automatically whenever needed.
